tablecodec/tablecodec.go

```diff
@@ -934,3 +938,190 @@ func GetTableIndexKeyRange(tableID, indexID int64) (startKey, endKey []byte) {
 	endKey = EncodeIndexSeekKey(tableID, indexID, []byte{255})
 	return
 }
+
+// GetIndexKeyBuf reuse or allocate buffer
+func GetIndexKeyBuf(buf []byte, defaultCap int) []byte {
+	if buf != nil {
+		return buf[:0]
+	}
+	return make([]byte, 0, defaultCap)
+}
+
+// GenIndexKey creates index keys using input buf
+func GenIndexKey(sc *stmtctx.StatementContext, tblInfo *model.TableInfo, idxInfo *model.IndexInfo,
+	idxPrefix kv.Key, indexedValues []types.Datum, h kv.Handle, buf []byte) (key []byte, distinct bool, err error) {
```

Comment:
Can we remove `idxPrefix`?

Reply:
The table id could be `physicalTableID` which is different from `TableInfo.ID` for partition table, `prefix` or `physicalTableID` is still needed as param.

Reply:
I think `physTblID` is better, the caller don't need to create idxPrefix.
In the amend index case, the `physTblID` can be extracted from the row key.

Reply:
Got

Reply:
A new method "GenIndexKeyUsingPrefix" is wrapped, `Index.Create` will use this method with the already encoded prefix byte array.  "GenIndexKey" has physical table id as input param and will do the prefix encode itself.